juju/juju
# Restart cmr workers after offer has been removed

There's scenarios which can break cross model relations deployments. This PR contains a fix for some of the issues.  In addition, investigating this found a problem with the unit agent permission checks which is also fixed. Finally, the key user facing errors out of state have been tweaked to use the term "saas" rather than "remote" as this matches what we show in status.

The key fixes:
1. consuming a terminated offer a second time
When an offer is removed, it is marked as terminated in the consuming model. A new offer can be consumed again, but if it has the same name as before, the terminated offer record was not replaced properly. To get the consuming model to notice the offer is consumed again, the old terminated one has to be marked as Dead and then removed.

Part of doing that fix resulted in some common "iterate over relations and leave scope" code pulled into a new method.

2. remote relations workers
Fixing  issue 1 fired events to the remote relations worker. But this worker was not properly noticing that the consumed application was for a different offer uuid (the name was the same). So extra logic was added to stop and start a new worker as needed. Also, when an offer is terminated, we need to immediately stop the status watcher.

3. Unit agent permissions
If a unit agent asks the Life or CharmModifiedVersion or CharmURL of its application, the unit agent permission checks were rejecting this. This isn't directly a cmr issue but is fixed here as a driveby.
Also, if an invalid unit tag is passed in to LeaveScope() surface that error rather than a permission one.

Still todo - when an offer is --force removed, not all internal cross model artefacts are cleaned up and the next time a new offer of the same name is used, the set up is not properly done and relations don't run all the hooks.

## QA steps

Create an offer
Consume the offer but do not relate to it
Remove the offer
Observe the saas application goes to terminated.
Consume the offer again (do not remove the terminated saas app first) and relate to it
Observe the relation hooks running

Previously consuming the offer a second time would not set up the workers to process the relation.

## Bug reference

Partial fix for https://bugs.launchpad.net/bugs/1940983
